content/articles/fcm-with-spring-and-angular/index.md

```diff
@@ -0,0 +1,500 @@
+           logger.info("Firebase application has been initialized")
+       }
+   } catch (e: IOException) {
+       logger.error(e.message)
+   }
+
+}
+```
+With all this, our Spring Boot application should be configured.
+
+### Creating a Firebase Cloud Messaging service
+
+To start, let’s create a service to send notifications and subscribe users to a topic. But first, let’s create a few model classes to represent a notification:
+
+```kotlin
+abstract class AppNotification(open val title: String, open val message: String)
+
+data class TopicNotification(val topic: String, override val title: String,
+                             override val message: String): AppNotification(title, message)
+
+data class DirectNotification(val target: String, override val title: String,
+                              override val message: String): AppNotification(title, message)
+```
+Using these classes, we can create two functions in our service to send direct and topic notifications respectively:
+
+```kotlin
+package me.john.amiscaray.services
+
+import org.springframework.stereotype.Service
+import com.google.Firebase.messaging.*
+import me.john.amiscaray.dtos.SubscriptionRequest
+import me.john.amiscaray.dtos.DirectNotification
+import me.john.amiscaray.dtos.TopicNotification
+
+
+@Service
+class FCMService {
+
+   fun sendNotificationToTarget(notification: DirectNotification){
+
+       val message = Message.builder()
+               .setWebpushConfig(
+                       WebpushConfig.builder()
+                               .setNotification(
+                                       WebpushNotification.builder()
+                                               .setTitle(notification.title)
+                                               .setBody(notification.message)
+                                               .setIcon("https://assets.mapquestapi.com/icon/v2/circle@2x.png")
+                                               .build()
+                               ).build()
+               )
+               .setToken(notification.target)
+               .build()
+       FirebaseMessaging.getInstance().sendAsync(message)
+
+   }
+
+   fun sendNotificationToTopic(notification: TopicNotification){
+
+       val message = Message.builder()
+               .setWebpushConfig(
+                       WebpushConfig.builder()
+                               .setNotification(
+                                       WebpushNotification.builder()
+                                               .setTitle(notification.title)
+                                               .setBody(notification.message)
+                                               .setIcon("https://assets.mapquestapi.com/icon/v2/incident@2x.png")
+                                               .build()
+                               ).build()
+               ).setTopic(notification.topic)
+               .build()
+
+       FirebaseMessaging.getInstance().sendAsync(message)
+
```

Comment:
Break down this code snippet into 2 smaller code snippets, with each containing one function. And, explain what each function does under the respective code snippet.

Also, add comments if possible.